### DISCUSSION TITLE
Update `HostFilteringOptions` based on K8s Ingresses

### DISCUSSION THREAD

**Discussion by pinkfloydx33** [dotnet/yarp#1824] [Q&A] [answered]
When using the Kubernetes controller, is there a way to have it reflect the hosts it detects on `Ingress` resources into the `HostFilteringOptions` used by the `HostFilteringMiddleware`?

We currently (without the controller) set the `AllowedHosts` based on a pre-determined list of host names. We are looking at moving to the Ingress Controller and would ideally like the `HostFilteringOptions` to be populated/updated in response to changes in `Ingress`/routes. That would play nicely with something like `external-dns` (which we also use) and allow us to have the downstream changes reflected in YARP without requiring us to update _it_ as well. 

At quick glance it doesn't look like there's anything built in that does this. I assume it would require a custom `IConfigureOptions<HostFilteringOptions>` as well as an `IOptionsChangeTokenSource<HostFilteringOptions>` for the options monitor to reflect any updates.

I have an idea how to implement the former by injecting `IProxyStateLookup`, looping over `GetRoutes` and extracting/parsing all the `Host` matches (sans Port),  but I'm honestly not sure how to implement the `IOptionsChangeTokenSource<>` to react to/trigger based on updates to the routes.

Any ideas? 

Thinking about this a bit more... I don't suppose this actually needs to be limited to the Kubernetes case. It feels like something that can be useful all around, opt-in of course

**Comment by adityamandaleeka** [comment, marked as answer]
Triage: this seems like it should be doable once https://github.com/microsoft/reverse-proxy/pull/1734 is merged. 

Essentially you would need to listen to the config change and look for hosts, then trigger the `IOptionsMonitor` which will update the `HostFilteringMiddleware` (assuming that's what you want).

**Comment by pinkfloydx33** [comment]
Right. I guess I was mainly trying to figure out what I'd use for the change token source. I guess I'd use a cancellation change token source and trigger it from the config notifier. Seems doable. Cool
